src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/SharedUtils.java

```diff
@@ -356,6 +357,22 @@ public static boolean isTrivial(FunctionDescriptor cDesc) {
                 .orElse(false);
     }
 
+    public interface Allocator {
+        default MemorySegment allocate(MemoryLayout layout) {
+            return allocate(layout.byteSize(), layout.byteAlignment());
+        }
+
+        default MemorySegment allocate(long size) {
+            return allocate(size, 1);
+        }
+
+        MemorySegment allocate(long size, long align);
+
+        static Allocator ofScope(NativeScope scope) {
+            return (size, align) -> scope.allocate(size, align).segment();
```

Comment:
IIUC, this segment won't have CLOSE mode, so when VaList.close() called, wouldn't that cause exception?

Reply:
Yes, it would cause an exception. It's the same when allocating from a NativeScope like normal. You don't have to call close() in that case.